Source/MLX/MLXArray+Indexing.swift

```diff
@@ -521,4 +353,1225 @@ extension MLXArray {
+    /// - ``MLXArrayIndex/stride(from:to:by:)``
+    public subscript(indices: MLXArrayIndex..., stream stream: StreamOrDevice = .default)
+        -> MLXArray
+    {
+        get {
+            // convert the e.g. Int, MLXArray, Range, etc. into operations
+            self[operations: indices.map { $0.mlxArrayIndexOperation }, stream: stream]
+        }
+        set {
+            self[operations: indices.map { $0.mlxArrayIndexOperation }, stream: stream] = newValue
+        }
+    }
+
+    /// General array indexing.
+    ///
+    /// See ``MLXArray/subscript(_:stream:)-375a0``
+    public subscript(indices: [MLXArrayIndex], stream stream: StreamOrDevice = .default)
+        -> MLXArray
+    {
+        get {
+            // convert the e.g. Int, MLXArray, Range, etc. into operations
+            self[operations: indices.map { $0.mlxArrayIndexOperation }, stream: stream]
+        }
+        set {
+            self[operations: indices.map { $0.mlxArrayIndexOperation }, stream: stream] = newValue
+        }
+    }
+
+}
+
+// MARK: - Support
+
+func countNonNewAxisOperations(_ operations: any Sequence<MLXArrayIndexOperation>) -> Int {
+    operations
+        .filter { !$0.isNewAxis }
+        .count
+}
+
+/// Replace `.ellipsis` operations with `.slice` operations for the expanded axes.
+///
+/// - Parameters:
+///   - shape: array shape
+///   - operations: operations
+/// - Returns: new operations with `.ellipsis` replaced
+func expandEllipsisOperations(shape: [Int32], operations: [MLXArrayIndexOperation])
+    -> [MLXArrayIndexOperation]
+{
+    // equivalent of mlx_expand_ellipsis
+
+    let ellipsisCount = operations.lazy.filter { $0.isEllipsis }.count
+    if ellipsisCount == 0 {
+        return operations
+    }
+    if ellipsisCount > 1 {
+        fatalError("multiple .ellipsis not allowed: \(operations)")
+    }
+
+    let prefix = operations.prefix { !$0.isEllipsis }
+    let suffix = operations.suffix(from: prefix.count + 1)
+
+    // inject full range slices in place of the ellipsis
+    let expandRange =
+        countNonNewAxisOperations(prefix) ..< (shape.count - countNonNewAxisOperations(suffix))
+    let expand = expandRange.map {
+        MLXArrayIndexOperation.slice(.init(start: 0, end: shape[$0], stride: 1))
+    }
+
+    return prefix + expand + suffix
+}
+
+// MARK: - index get
+
+/// Single operation MLXArray subscript get.
+///
+/// This provides optimized operations for simple forms of subscripts.
+///
+/// - Parameters:
+///   - src: input array
+///   - operation: single index operation
+/// - Returns: MLXArray get result
+func getItem(src: MLXArray, operation: MLXArrayIndexOperation, stream: StreamOrDevice = .default)
+    -> MLXArray
+{
+    switch operation {
+    case .ellipsis:
+        return src
+
+    case .newAxis:
+        return src.expandedDimensions(axis: 0, stream: stream)
+
+    case .index(let index):
+        return src.take(src.resolve(index: index, axis: 0), axis: 0, stream: stream)
```

Comment:
This is not an actionable comment as it is the same in C++/Python but we do have an inconsistency now when using a single index in the first axis or any other axis.

```python
x[0]       # copies
x[..., 0]  # doesn't copy
```

Don't know if we should change that 🤔. @jagrit06 and @awni wdyt?

Reply:
The single array case is the same:

```
    case .array(let indices):
        return src.take(indices, axis: 0, stream: stream)
```

Is this a `take` vs `gather`/`slice` issue?

Reply:
Yeah exactly. The single array has to copy because we may want any "row" (and any number of rows) in that array.